dask/array/core.py

```diff
@@ -1497,6 +1497,16 @@ def __xor__(self, other):
     def __rxor__(self, other):
         return elemwise(operator.xor, other, self)
 
+    def __matmul__(self, other):
+        return tensordot(self, other, axes=((self.ndim - 1,),
+                                            (other.ndim - 2,)))
+
+    def __rmatmul__(self, other):
+        if not hasattr(other, 'ndim'):
+            other = np.asarray(other)  # account for array-like on LHS
+        return tensordot(other, self, axes=((other.ndim - 1,),
+                                            (self.ndim - 2,)))
```

Comment:
> imho dask.array is in a different category, because it is designed to mimic dense NumPy arrays

There has been a slight shift recently.  Dask.array supports any container array that satisfies the numpy ndarray interface.  This now includes the nascent `sparse.COO` object here: https://github.com/mrocklin/sparse .  It could also include other ndarray objects in the future if they develop.

However, I agree with you that this is so far a very rare case.  I don't think that this is likely to be a common case and so shouldn't hold back progress if there aren't ways to conveniently support it.